File: scripts/python_ref/embed2d.py
#!/usr/bin/env python3
"""embed2d.py — map a 1D boolean stream into a 2D field for spatial analysis.

Mapping:
- choose width W
- bit index t -> (x=t%W, y=t//W)

Provides:
- tile extraction
- box-counting sketch for fractal-ish summaries
"""

from __future__ import annotations

from typing import List, Tuple


def embed(bits: List[int], W: int) -> List[List[int]]:
    if W <= 0:
        raise ValueError("W must be > 0")
    H = (len(bits) + W - 1) // W
    grid = [[0] * W for _ in range(H)]
    for t, b in enumerate(bits):
        grid[t // W][t % W] = b & 1
    return grid


def box_count(grid: List[List[int]], box: int) -> int:
    """Count boxes (box x box) that contain at least one 1."""
    if box <= 0:
        return 0
    H = len(grid)
    W = len(grid[0]) if H else 0
    cnt = 0
    for y0 in range(0, H, box):
        for x0 in range(0, W, box):
            hit = 0
            for y in range(y0, min(H, y0 + box)):
                row = grid[y]
                for x in range(x0, min(W, x0 + box)):
                    if row[x]:
                        hit = 1
                        break
                if hit:
                    break
            cnt += hit
    return cnt


def multiscale_box_counts(grid: List[List[int]], boxes: List[int]) -> List[Tuple[int,int]]:
    return [(b, box_count(grid, b)) for b in boxes]

File: internal/signal/embed2d/embed2d.go
package embed2d

// 1D boolean stream -> 2D embedding + simple box-counting sketch.

func Embed(bits []uint8, width int) [][]uint8 {
	if width <= 0 {
		width = 1
	}
	h := (len(bits) + width - 1) / width
	grid := make([][]uint8, h)
	for y := 0; y < h; y++ {
		grid[y] = make([]uint8, width)
	}
	for t, b := range bits {
		y := t / width
		x := t % width
		grid[y][x] = b & 1
	}
	return grid
}

func BoxCount(grid [][]uint8, box int) int {
	if box <= 0 || len(grid) == 0 {
		return 0
	}
	h := len(grid)
	w := len(grid[0])
	cnt := 0
	for y0 := 0; y0 < h; y0 += box {
		for x0 := 0; x0 < w; x0 += box {
			hit := false
			for y := y0; y < h && y < y0+box && !hit; y++ {
				row := grid[y]
				for x := x0; x < w && x < x0+box; x++ {
					if row[x] == 1 {
						hit = true
						break
					}
				}
			}
			if hit {
				cnt++
			}
		}
	}
	return cnt
}

type BoxCountPair struct {
	Box   int
	Count int
}

func MultiScaleBoxCounts(grid [][]uint8, boxes []int) []BoxCountPair {
	out := make([]BoxCountPair, 0, len(boxes))
	for _, b := range boxes {
		out = append(out, BoxCountPair{Box: b, Count: BoxCount(grid, b)})
	}
	return out
}
